### PR TITLE
Code permalink preview

### LINKED ISSUE #56: Code line preview
When pasting links to code parts it would be really awesome if the code preview could be rendered like on GitHub:

`https://github.com/nextcloud/integration_github/blob/3a3b259f8bae42d432f09b851be31c9c186e0e7b/lib/Reference/GithubReferenceProvider.php?rgh-link-date=2022-10-20T07%3A14%3A23Z#L37-L41`

https://github.com/nextcloud/integration_github/blob/3a3b259f8bae42d432f09b851be31c9c186e0e7b/lib/Reference/GithubReferenceProvider.php?rgh-link-date=2022-10-20T07%3A14%3A23Z#L37-L41

### PR DESCRIPTION
**Pull request by julien-nc**
closes #56

Here is a basic implementation of a reference widget for the code permalink previews like mentioned in #56.

As far as I've seen, it is not possible to get a set of lines with the GitHub REST API but we can get a full file as it was at a specific commit. We then only keep the lines selected in the link fragment/hash-parameter.

Here is a GitHub rendered one:
https://github.com/nextcloud/integration_github/blob/3a3b259f8bae42d432f09b851be31c9c186e0e7b/lib/Reference/GithubReferenceProvider.php#L79-L82

And it looks like that for the moment (in Talk):
image

It's using [vue-highlightjs](https://github.com/metachris/vue-highlightjs) (using highlight.js itself)
The `github-dark-dimmed` and `github` highlight.js themes are exactly what we needed :grin:.
There are other libraries allowing to add line numbers but it's usually hard to start from something > 1. Thankfully prepending the line number at the beginning of every line does not disturb the highlight.js syntax parser.

Remaining issues:
* [x] I didn't find a way to conditionally load a theme (done by loading a specific css file) depending on if Nextcloud is in dark mode or not

### CODE REVIEW

File: src/views/GithubCodePermalinkReferenceWidget.vue [hunk 1]
```diff
@@ -181,6 +184,28 @@ export default {
 }
 </script>
 
+<!--style lang="scss">
```
Comment (juliushaertl) [inline src/views/GithubCodePermalinkReferenceWidget.vue:187]: @julien-nc This works fine for me as soon as I comment out the code in beforeMount

Comment (juliushaertl) [reply src/views/GithubCodePermalinkReferenceWidget.vue:187]: Oh wait no, was mislead by browser cache

File: src/views/GithubCodePermalinkReferenceWidget.vue [hunk 2]
```diff
@@ -181,6 +184,28 @@ export default {
...
+// does not work :-)
+body[data-theme-default] {
+	@media (prefers-color-scheme: dark) {
+		@import 'highlight.js/styles/github-dark-dimmed.css';
```
Comment (juliushaertl) [inline src/views/GithubCodePermalinkReferenceWidget.vue:191]: Does not seem to work as the import matches the criteria from https://sass-lang.com/documentation/at-rules/import#plain-css-imports

Changing to this seems to do the trick though:

```suggestion
		@import 'highlight.js/styles/github-dark-dimmed';
```

Comment (julien-nc) [reply src/views/GithubCodePermalinkReferenceWidget.vue]: :pray: Thanks a lot. All good now.

File: css/github.css [hunk 3]
```diff
@@ -0,0 +1,10 @@
+pre code.hljs{display:block;overflow-x:auto;padding:1em}code.hljs{padding:3px 5px}/*!
```
Comment (juliushaertl) [inline css/github.css:1]: Those css files seem to be unused now ;)

**Review by juliushaertl — APPROVED** [review]
Maybe some more cleanup needed on the CSS files, but otherwise LGTM 👍 

Awesome work 🚀

**Comment by julien-nc** [comment]
Thanks a lot for the lucid review! :blue_heart:

**Pull request merged**